Da monitoring service

jfldde opened the pull request:
# Description

- Monitor txs send in `spawn_da_queue`
- Monitor chain re-org and resubmit txs
- Add `DaRpc` server/client
- Add cpfp fee bumping endpoint: `da_bumpFeeCpfp` which takes an Optional<Txid> and a new fee_rate. Bumps latest TX in chain if txid is None
- Add monitoring endpoints : `da_getPendingTransactions`, `da_getTxStatus` and `da_getLastMonitoredTx`
- Revive TX chain on `MonitoringService` initialization and keep as single source of truth for `prev_utxo` and `pending_txs`

## Linked Issues
- Related to #892 

## Tests
- Add a re-org test asserting that txs are properly re-submitted and commitments are part of the updated height post re-org
- Add a cpfp fee bumping test that assert that TX chain is followed.

## Missing
- Add test again full mempool
- Add RBF fee bumping
- Add auto fee bumping trigger if needed

jfldde committed:
With monitoring

jfldde committed:
Rename failed to evicted

jfldde committed:
Cleanup

jfldde committed:
With config

jfldde committed:
Bitcoin re-org test

jfldde committed:
Resubmit

jfldde committed:
Cleanup

jfldde committed:
Keep track of initial block height on monitored txs

eyusufatik commented on crates/bitcoin-da/src/service.rs:
```diff
@@ -242,8 +264,11 @@ impl BitcoinService {
                                                 let _ = request.notify.send(Ok(tx_id));
+
+                                                if let Err(e) = self.monitoring.monitor_transaction_chain(txids).await {
+                                                    error!(?e, "Failed to monitored tx chain");
```
monitor*

eyusufatik commented on crates/bitcoin-da/src/service.rs:
```diff
@@ -358,7 +383,7 @@ impl BitcoinService {
         da_data: DaData,
         fee_sat_per_vbyte: u64,
-    ) -> Result<TxWithId, anyhow::Error> {
+    ) -> Result<(Vec<Txid>, TxWithId), anyhow::Error> {
```
Just make it Vec<TxWithId>?

jfldde replied:
You mean a Vec of `Txid` or `TxWithId` ? 
We need the full Transaction from `TxWithId` to build the `prev_utxo`.
We could hold a Vec of Transaction but we still need to call `gettransaction <txid>` in the monitoring service to fetch the number of confirmations and to assert that it's held in mempool

eyusufatik replied:
Vec of TxWithId

jfldde replied:
The TxWithId is used to build the next prev_utxo whereas the Vec<Txid> is used to monitor all the txs.
We don't really need to hold the Transaction itself alongside them since we need to do a gettransaction call anyway to check confirmations.

eyusufatik commented on crates/bitcoin-da/src/monitoring.rs:
```diff
@@ -0,0 +1,502 @@
+// Todo pass down lower value for test
+// const DEFAULT_CHECK_INTERVAL: Duration = Duration::from_secs(60);
+const DEFAULT_CHECK_INTERVAL: Duration = Duration::from_millis(10);
+const DEFAULT_HISTORY_LIMIT: usize = 1_000; // Keep track of last 1k txs
```
putting size limitation on the history is probably helpful as well.

jfldde replied:
Other than the DEFAULT_HISTORY_LIMIT you mean ?

eyusufatik replied:
yeah the size of all txs in history, might be unnecessary though.

but for the prover for example

assume at high traffic for a long time 

400kb for 2 tx groups (commit + reveal)

totals at 195 MB

hmm, that's fine actually

or is the history just txid?

jfldde replied:
Sounded good and future proof, added it

eyusufatik commented on crates/bitcoin-da/src/monitoring.rs:
```diff
@@ -0,0 +1,502 @@
+    Replaced {
+        by_txid: Txid,
+    },
+    Evicted,
```
what is evicted

jfldde replied:
That would be the TX status if it ever goes from `Pending` to not found in mempool. 
Could be a good trigger for us to bump fee as eviction can happen in time of congestion when a mempool gets full.

eyusufatik commented on crates/bitcoin-da/src/service.rs:
can put monitoring service config under bitcoin service config, that way we can set it through our config toml's

jfldde committed:
Fix grammar

jfldde committed:
Add monitoring to BitcoinServiceConfig

jfldde committed:
RPCs server/client

jfldde committed:
Use only concrete BitcoinService type
Will be mostly used for mempool monitoring and fee bumping and won't be needed on MockDa

jfldde committed:
Simplify and add get_tx_status method

jfldde committed:
Remove unused metrics

jfldde committed:
Merge branch 'nightly' of github.com:chainwayxyz/citrea into tx-monitoring-service

jfldde committed:
Bump e2e rev

jfldde committed:
Merge branch 'nightly' of github.com:chainwayxyz/citrea into tx-monitoring-service

jfldde committed:
Bump fee via cpfp

jfldde committed:
Merge branch 'nightly' into tx-monitoring-service

jfldde committed:
Lint

jfldde committed:
Remove unused import

jfldde committed:
Bump citrea-e2e

jfldde committed:
Remove udeps

jfldde committed:
Lift up bump_fee_cpfp

jfldde committed:
Clearer comment

jfldde committed:
Make MonitoringConfig optional

jfldde committed:
Revert empty line change

eyusufatik commented on crates/bitcoin-da/src/service.rs:
```diff
@@ -183,13 +215,16 @@ impl BitcoinService {
         self: Arc<Self>,
         mut rx: UnboundedReceiver<Option<SenderWithNotifier<TxidWrapper>>>,
     ) {
+        self.monitoring.clone().spawn();
+
         tokio::spawn(async move {
             let mut prev_utxo = match self.get_prev_utxo().await {
                 Ok(Some(prev_utxo)) => Some(prev_utxo),
```
should we get the prev_utxo from the monitoring service now?

and update it from there so if tx is bumped, we use an output from the bump tx

eyusufatik commented on crates/bitcoin-da/src/service.rs:
```diff
@@ -357,17 +395,25 @@ impl BitcoinService {
 
+        // If Utxo has already been spent, fallback to get_prev_utxo
+        // Would happen in case of cpfp spending reveal output
+        if let Some(prev) = &prev_utxo {
+            if !utxos.iter().any(|utxo| utxo == prev) {
+                prev_utxo = self.get_prev_utxo().await?;
+            }
+        }
```
(referring to my last comment) and if we do it this way, now we don't need to search in the utxo list, the monitoring chain becomes the source of truth for `prev_utxo`

jfldde replied:
Sounds good, I'll amend

eyusufatik commented on crates/bitcoin-da/src/service.rs:
```diff
@@ -582,6 +628,128 @@ impl BitcoinService {
+            // else take any available utxo
+            for (idx, utxo) in parent_tx.output.iter().enumerate().skip(1) {
+                if total_input > required_fee {
+                    break;
+                }
+                child_tx.input.push(create_tx_input(OutPoint {
+                    txid: parent_txid,
+                    vout: idx as u32,
+                }));
+                total_input += utxo.value;
+            }
```
as you keep adding this inputs, you have to recalculate the fee needed to pay since the size of the child increases :)

jfldde committed:
Recalculate child fee on each additional input

jfldde committed:
Move break condition since we removed first prev_tx utxo loop

jfldde committed:
Revive monitoring service from mempool

jfldde committed:
Keep finalized confirmations

jfldde committed:
Merge branch 'nightly' into tx-monitoring-service

eyusufatik commented on crates/bitcoin-da/src/service.rs:
```diff
@@ -684,45 +651,33 @@ impl BitcoinService {
                 }));
 
                 total_input += utxo.amount;
+                required_fee = calculate_child_fee(&child_tx);
```
wouldn't you need total_input >= required_fee + output amount? or smth like that?

jfldde committed:
Keep at least as much output as parent tx

jfldde committed:
Refactor get_prev_utxo to use monitoring service

jfldde committed:
Merge branch 'nightly' of github.com:chainwayxyz/citrea into tx-monitoring-service

jfldde committed:
Add max history size

jfldde committed:
Cleanup monitoring init

jfldde committed:
Fix test flakiness, give time to monitoring service

jfldde committed:
Bump bitcoincore-rpc

eyusufatik commented on Cargo.toml:
```diff
@@ -156,4 +156,4 @@ tower = { version = "0.4.13", features = ["full"] }
 
 [patch.crates-io]
-bitcoincore-rpc = { version = "0.18.0", git = "https://github.com/chainwayxyz/rust-bitcoincore-rpc.git", rev = "ede8097" }
+bitcoincore-rpc = { version = "0.18.0", git = "https://github.com/chainwayxyz/rust-bitcoincore-rpc.git", rev = "5ce1bed" }
```
if this closes the bitcoin core rpc bump issue pls add it to the pr sidebar

jfldde replied:
This is not addressing issue #1364 , only bumping to include https://github.com/chainwayxyz/rust-bitcoincore-rpc/pull/6 RPC typing fix

eyusufatik commented on bin/citrea/tests/bitcoin_e2e/bitcoin_test.rs:
```diff
@@ -0,0 +1,366 @@
+            .client
+            .http_client()
+            .da_get_pending_transactions()
+            .await?;
```
assert fee is really bumped to where it should be

jfldde committed:
Merge branch 'nightly' of github.com:chainwayxyz/citrea into tx-monitoring-service

rakanalh commented on bin/citrea/src/rollup/bitcoin.rs:
```diff
@@ -162,6 +166,7 @@ impl RollupBlueprint for BitcoinRollup {
         if require_wallet_check {
             // run only for sequencer and prover
             Arc::clone(&service).spawn_da_queue(rx);
+            service.monitoring.clone().spawn().await?;
```
Please think about how we can use a cancellation token with `ctrl_c` from tokio to support future cancellation for both the da queue service and the monitoring service.

jfldde replied:
[efc1ba1](https://github.com/chainwayxyz/citrea/pull/1407/commits/efc1ba1ffe9b49505ceb610d89a54f7dca03d406) Sharing the task manager and handling it via the main loop select

rakanalh commented on crates/bitcoin-da/src/monitoring.rs:
```diff
@@ -0,0 +1,510 @@
+    pub async fn spawn(self: Arc<Self>) -> Result<()> {
+        self.init().await?;
+
+        tokio::spawn(async move {
```
Accept a cancellation token as a param to `spawn` and use it here to stop if the cancellation token is triggered.

rakanalh commented on crates/bitcoin-da/src/monitoring.rs:
```diff
@@ -0,0 +1,510 @@
+    //         .iter()
+    //         .map(|(txid, tx)| (*txid, tx.clone()))
+    //         .collect()
+    // }
```
Remove?

jfldde committed:
Cleanup anyhow

jfldde committed:
Fix test

jfldde committed:
Remove dead code

jfldde committed:
Merge branch 'nightly' into tx-monitoring-service

jfldde committed:
Use TaskManager for DaService and various cleanup

jfldde committed:
Merge branch 'da-task-manager' of github.com:chainwayxyz/citrea into tx-monitoring-service

jfldde committed:
Lint

jfldde committed:
Fix init test

jfldde committed:
Merge branch 'nightly' of github.com:chainwayxyz/citrea into tx-monitoring-service

jfldde committed:
Fmt select loop

jfldde committed:
Use psbt and test fee calculation

jfldde committed:
Merge branch 'nightly' into tx-monitoring-service

jfldde committed:
Lint

jfldde committed:
fmt

jfldde committed:
Keep track of MonitoredTxKind and prevent bumping commit tx
